src/predictor/gpu_predictor.cu

```diff
@@ -323,20 +324,26 @@ class GPUPredictor : public xgboost::Predictor {
                 h_nodes.begin() + h_tree_segments[tree_idx - tree_begin]);
     }
 
-    size_t i_batch = 0;
-
+    std::vector<size_t> batch_offsets(devices_.Size(), 0);
+    GPUDistribution empty_distribution;
     for (const auto &batch : dmat->GetRowBatches()) {
-      CHECK_EQ(i_batch, 0) << "External memory not supported";
       // out_preds have been resharded and resized in InitOutPredictions()
       batch.offset.Reshard(GPUDistribution::Overlap(devices_, 1));
       std::vector<size_t> device_offsets;
       DeviceOffsets(batch.offset, &device_offsets);
       batch.data.Reshard(GPUDistribution::Explicit(devices_, device_offsets));
       dh::ExecuteIndexShards(&shards_, [&](int idx, DeviceShard& shard) {
-        shard.PredictInternal(batch, dmat->Info(), out_preds, model,
+        shard.PredictInternal(batch, dmat->Info(), out_preds, batch_offsets[idx], model,
                               h_tree_segments, h_nodes, tree_begin, tree_end);
       });
-      i_batch++;
+      for (int i = 0; i < devices_.Size(); i++) {
+        if (batch.offset.Devices().Contains(i)) {
+          size_t num_rows = batch.offset.DeviceSize(i) - 1;
+          batch_offsets[i] += num_rows * model.param.num_output_group;
+        }
+      }
+      batch.offset.Reshard(empty_distribution);
```

Comment:
i'm wondering why we need this. is it to make [this](https://github.com/dmlc/xgboost/blob/master/src/common/host_device_vector.cu#L363) check happy?

more fundamentally, why is that check present? is it not plausible to first shard a device vector based on a gpu distribution config, only to later reshard with another distribution with a different configuration? for my own edification, what is forcing one of the distribution to be always non-empty?

Reply:
@sriramch I think it was added to prevent accidental data movement.

Reply:
can you please expand? wouldn't reshard recreate a new distribution (reinitialize the shards) based on the new distribution config, thus effectively wiping out the old one? when would there be a `data movement`?

Reply:
If memory serves I think it was added by @canonizer .  And it's to make sure if one want to move data from one device to another, or to gather/scatter , it has to be done explicitly.  We might be doing this accidentally due to wrong parameters, data granularity etc.